File: server/module/module-erp/erp-model/src/response/erp_purchase_order.rs
use chrono::NaiveDateTime;
use serde::{Serialize, Deserialize};
use utoipa::ToSchema;
use serde_with::{serde_as, DisplayFromStr};
use common::formatter::string_date_time::StringDateTime;

use crate::response::{erp_purchase_order_attachment::{ErpPurchaseOrderAttachmentBaseResponse, ErpPurchaseOrderAttachmentInfoResponse}, erp_purchase_order_detail::{ErpPurchaseOrderDetailBaseResponse, ErpPurchaseOrderDetailInfoResponse}};

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpPurchaseOrderResponse {
    
    pub id: i64, // 采购订单ID
    
    pub order_number: i64, // 订单编号
    
    pub supplier_id: i64, // 供应商ID
    
    pub user_id: i64, // 用户ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub purchase_date: NaiveDateTime, // 采购日期
    
    pub total_amount: i64, // 总金额
    
    pub order_status: i8, // 订单状态
    
    pub discount_rate: Option<i64>, // 优惠率（百分比，1000表示10.00%）
    
    pub settlement_account_id: Option<i64>, // 结算账户ID
    
    pub deposit: Option<i64>, // 定金
    
    pub remarks: Option<String>, // 备注
    
    pub department_code: String, // 部门编码
    
    pub department_id: i64, // 部门ID
    
    pub creator: Option<i64>, // 创建者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub create_time: NaiveDateTime, // 创建时间
    
    pub updater: Option<i64>, // 更新者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub update_time: NaiveDateTime, // 更新时间
    
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpPurchaseOrderPageResponse {
    
    pub id: i64, // 采购订单ID
    
    pub order_number: i64, // 订单编号
    
    pub supplier_id: i64, // 供应商ID
    
    pub user_id: i64, // 用户ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub purchase_date: NaiveDateTime, // 采购日期
    
    pub total_amount: i64, // 总金额
    
    pub order_status: i8, // 订单状态
    
    pub discount_rate: Option<i64>, // 优惠率（百分比，1000表示10.00%）
    
    pub settlement_account_id: Option<i64>, // 结算账户ID
    
    pub deposit: Option<i64>, // 定金
    
    pub remarks: Option<String>, // 备注
    
    pub department_code: String, // 部门编码
    
    pub department_id: i64, // 部门ID
    
    pub creator: Option<i64>, // 创建者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub create_time: NaiveDateTime, // 创建时间
    
    pub updater: Option<i64>, // 更新者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub update_time: NaiveDateTime, // 更新时间

    /****************** 信息 ******************/

    pub supplier_name: Option<String>, // 供应商名

    pub settlement_account_name: Option<String>, // 结算账户名
    
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpPurchaseOrderBaseResponse {
    
    pub id: i64, // 采购订单ID
    
    pub order_number: i64, // 订单编号
    
    pub supplier_id: i64, // 供应商ID
    
    pub user_id: i64, // 用户ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub purchase_date: NaiveDateTime, // 采购日期
    
    pub total_amount: i64, // 总金额
    
    pub order_status: i8, // 订单状态
    
    pub discount_rate: Option<i64>, // 优惠率（百分比，1000表示10.00%）
    
    pub settlement_account_id: Option<i64>, // 结算账户ID
    
    pub deposit: Option<i64>, // 定金
    
    pub remarks: Option<String>, // 备注

    pub purchase_products: Vec<ErpPurchaseOrderDetailBaseResponse>, // 采购的产品列表

    pub purchase_attachment: Vec<ErpPurchaseOrderAttachmentBaseResponse>, // 采购的附件列表
    
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpPurchaseOrderInfoResponse {
    
    pub id: i64, // 采购订单ID
    
    pub order_number: i64, // 订单编号
    
    pub supplier_id: i64, // 供应商ID
    
    pub user_id: i64, // 用户ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub purchase_date: NaiveDateTime, // 采购日期
    
    pub total_amount: i64, // 总金额
    
    pub order_status: i8, // 订单状态
    
    pub discount_rate: Option<i64>, // 优惠率（百分比，1000表示10.00%）
    
    pub settlement_account_id: Option<i64>, // 结算账户ID
    
    pub deposit: Option<i64>, // 定金
    
    pub remarks: Option<String>, // 备注

    pub department_code: String, // 部门编码
    
    pub department_id: i64, // 部门ID
    
    pub creator: Option<i64>, // 创建者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub create_time: NaiveDateTime, // 创建时间
    
    pub updater: Option<i64>, // 更新者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub update_time: NaiveDateTime, // 更新时间

    /****************** 信息 ******************/

    pub supplier_name: Option<String>, // 供应商名

    pub settlement_account_name: Option<String>, // 结算账户名

    pub purchase_products: Vec<ErpPurchaseOrderDetailInfoResponse>, // 采购的产品列表

    pub purchase_attachment: Vec<ErpPurchaseOrderAttachmentInfoResponse>, // 采购的附件列表
    
}

File: admin/src/api/erp_purchase_order.ts
import { PaginatedRequest, PaginatedResponse } from '@/base/page';
import { api } from '@/utils/request';
import { ErpPurchaseOrderDetailBaseResponse, ErpPurchaseOrderDetailInfoResponse, ErpPurchaseOrderDetailRequest } from './erp_purchase_order_detail';
import { ErpPurchaseOrderAttachmentBaseResponse, ErpPurchaseOrderAttachmentInfoResponse, ErpPurchaseOrderAttachmentRequest } from './erp_purchase_order_attachment';

const apis = {
  create: '/erp/erp_purchase_order/create', // 新增
  update: '/erp/erp_purchase_order/update', // 修改
  delete: '/erp/erp_purchase_order/delete', // 删除
  get: '/erp/erp_purchase_order/get', // 单条查询
  list: '/erp/erp_purchase_order/list', // 列表查询
  list_supplier: '/erp/erp_purchase_order/list_supplier', // 列表查询
  page: '/erp/erp_purchase_order/page', // 分页查询
  get_detail: '/erp/erp_purchase_order/get_detail', // 单条查询订单详情
  get_info: '/erp/erp_purchase_order/get_info', // 单条查询订单所有信息
  received: '/erp/erp_purchase_order/received', // 已收货
  cancel: '/erp/erp_purchase_order/cancel', // 取消订单
  page_received: '/erp/erp_purchase_order/page_received', // 分页查询待入库
}

export interface ErpPurchaseOrderRequest {
  id?: number; // 采购订单ID
  supplier_id: number; // 供应商ID
  purchase_date: string; // 采购日期
  total_amount: number; // 总金额
  discount_rate?: number; // 优惠率（百分比，1000表示10.00%）
  settlement_account_id?: number; // 结算账户ID
  deposit?: number; // 定金
  remarks?: string; // 备注

  purchase_products: ErpPurchaseOrderDetailRequest[]; // 采购的产品列表
  purchase_attachment: ErpPurchaseOrderAttachmentRequest[]; // 采购的附件列表
}

export interface ErpPurchaseOrderResponse {
  id: number; // 采购订单ID
  order_number: string; // 订单编号
  supplier_id: number; // 供应商ID
  user_id: number; // 用户ID
  purchase_date: string; // 采购日期
  total_amount: number; // 总金额
  order_status: number; // 订单状态 (0=pending, 1=completed, 2=cancelled)
  discount_rate: number; // 优惠率（百分比，1000表示10.00%）
  settlement_account_id: number; // 结算账户ID
  deposit: number; // 定金
  remarks: string; // 备注
  department_code: string; // 部门编码
  department_id: number; // 部门ID
  creator: number; // 创建者ID
  create_time: string; // 创建时间
  updater: number; // 更新者ID
  update_time: string; // 更新时间

  supplier_name: string; // 供应商名
  settlement_account_name: string; // 结算账户名
}

export interface ErpPurchaseOrderBaseResponse {
  id: number; // 采购订单ID
  order_number: string; // 订单编号
  supplier_id: number; // 供应商ID
  user_id: number; // 用户ID
  purchase_date: string; // 采购日期
  total_amount: number; // 总金额
  order_status: number; // 订单状态 (0=pending, 1=completed, 2=cancelled)
  discount_rate: number; // 优惠率（百分比，1000表示10.00%）
  settlement_account_id: number; // 结算账户ID
  deposit: number; // 定金
  remarks: string; // 备注

  purchase_products: ErpPurchaseOrderDetailBaseResponse[]; // 采购的产品列表
  purchase_attachment: ErpPurchaseOrderAttachmentBaseResponse[]; // 采购的附件列表
}

export interface ErpPurchaseOrderInfoResponse {
  id: number; // 采购订单ID
  order_number: string; // 订单编号
  supplier_id: number; // 供应商ID
  user_id: number; // 用户ID
  purchase_date: string; // 采购日期
  total_amount: number; // 总金额
  order_status: number; // 订单状态 (0=pending, 1=completed, 2=cancelled)
  discount_rate: number; // 优惠率（百分比，1000表示10.00%）
  settlement_account_id: number; // 结算账户ID
  deposit: number; // 定金
  remarks: string; // 备注
  department_code: string; // 部门编码
  department_id: number; // 部门ID
  creator: number; // 创建者ID
  create_time: string; // 创建时间
  updater: number; // 更新者ID
  update_time: string; // 更新时间

  supplier_name: string; // 供应商名
  settlement_account_name: string; // 结算账户名
  purchase_products: ErpPurchaseOrderDetailInfoResponse[]; // 采购的产品列表
  purchase_attachment: ErpPurchaseOrderAttachmentInfoResponse[]; // 采购的附件列表
}

export interface ErpPurchaseOrderQueryCondition extends PaginatedRequest {

}

export const createErpPurchaseOrder = (erp_purchase_order: ErpPurchaseOrderRequest): Promise<number> => {
  return api.post<number>(apis.create, erp_purchase_order);
}

export const updateErpPurchaseOrder = (erp_purchase_order: ErpPurchaseOrderRequest): Promise<void> => {
  return api.post<void>(apis.update, erp_purchase_order);
}

export const deleteErpPurchaseOrder = (id: number): Promise<void> => {
  return api.post<void>(`${apis.delete}/${id}`);
}

export const getErpPurchaseOrder = (id: number): Promise<ErpPurchaseOrderResponse> => {
  return api.get<ErpPurchaseOrderResponse>(`${apis.get}/${id}`);
}

export const listErpPurchaseOrder = (): Promise<Array<ErpPurchaseOrderResponse>> => {
  return api.get<Array<ErpPurchaseOrderResponse>>(apis.list);
}

export const pageErpPurchaseOrder = (condition: ErpPurchaseOrderQueryCondition): Promise<PaginatedResponse<ErpPurchaseOrderResponse>> => {
  return api.get<PaginatedResponse<ErpPurchaseOrderResponse>>(apis.page, condition);
}

export const getErpPurchaseOrderBase = (id: number): Promise<ErpPurchaseOrderBaseResponse> => {
  return api.get<ErpPurchaseOrderBaseResponse>(`${apis.get_detail}/${id}`);
}

export const getErpPurchaseOrderInfo = (id: number): Promise<ErpPurchaseOrderInfoResponse> => {
  return api.get<ErpPurchaseOrderInfoResponse>(`${apis.get_info}/${id}`);
}

export const receivedErpPurchaseOrder = (id: number): Promise<void> => {
  return api.post<void>(`${apis.received}/${id}`);
}

export const cancelErpPurchaseOrder = (id: number): Promise<void> => {
  return api.post<void>(`${apis.cancel}/${id}`);
}

export const pageReceivedErpPurchaseOrder = (condition: ErpPurchaseOrderQueryCondition): Promise<PaginatedResponse<ErpPurchaseOrderResponse>> => {
  return api.get<PaginatedResponse<ErpPurchaseOrderResponse>>(apis.page_received, condition);
}

export const listSupplierErpPurchaseOrder = (supplier_id: number): Promise<Array<ErpPurchaseOrderResponse>> => {
  return api.get<Array<ErpPurchaseOrderResponse>>(`${apis.list_supplier}/${supplier_id}`);
}
